src/compiler/collections.ts

```diff
@@ -0,0 +1,596 @@
+// NumberMap, StringMap, and StringSet shims
+/* @internal */
+namespace ts {
+    // The global Map object. This may not be available, so we must test for it.
+    declare const Map: { new<K, V>(pairs?: [K, V][]): Map<K, V> } | undefined;
+    const usingNativeMaps = typeof Map !== "undefined";
+    // tslint:disable-next-line:no-in-operator
+    const fullyFeaturedMaps = usingNativeMaps && "keys" in Map.prototype && "values" in Map.prototype && "entries" in Map.prototype;
+
+    /** Extra Map methods that may not be available, so we must provide fallbacks. */
+    interface ES6Map<K, V> extends Map<K, V> {
+        keys(): Iterator<K>;
+        values(): Iterator<V>;
+        entries(): Iterator<[K, V]>;
+    }
+
+    /** Simplified ES6 Iterator interface. */
+    interface Iterator<T> {
+        next(): { value: T, done: false } | { value: never, done: true };
+    }
+
+    /**
+     * Provides Map-like functionality for ES5 runtimes.
+     * This is intentionally *not* a full Map shim, and doesn't provide iterators (which aren't available for IE Maps anyway).
+     * We can only efficiently support strings and number keys, and iteration will always yield stringified keys.
+     */
+    class ShimMap<K extends string | number, V> implements Map<K, V> {
+        private data = createDictionaryModeObject<V>();
+
+        /*
+        So long as `K extends string | number`, we can cast `key as string` and insert it into the map.
+        However, `forEach` will iterate over strings because values are stringified before being put in the map.
+        */
+
+        constructor(pairs?: [K, V][]) {
+            if (pairs) {
+                for (const [key, value] of pairs) {
+                    this.data[key as string] = value;
+                }
+            }
+        }
+
+        clear(): void {
+            this.data = createDictionaryModeObject<V>();
+        }
+
+        delete(key: K): void {
+            delete this.data[key as string];
+        }
+
+        get(key: K): V {
+            return this.data[key as string];
+        }
+
+        has(key: K): boolean {
+            // tslint:disable-next-line:no-in-operator
+            return (key as string) in this.data;
+        }
+
+        set(key: K, value: V): void {
+            this.data[key as string] = value;
+        }
+
+        forEach(action: (value: V, key: string) => void): void {
+            for (const key in this.data) {
+                action(this.data[key], key);
+            }
+        }
+    }
+
+    /**
+     * In runtimes without Maps, this is implemented using an object.
+     * `pairs` is an optional list of entries to add to the new map.
+     */
+    export const StringMap: { new<T>(pairs?: [string, T][]): Map<string, T>; } = usingNativeMaps ? Map : ShimMap;

```

Comment:
I'd love for us to be able to just call this `Map` instead of `StringMap` or `NumberMap`, but its a bit messy to avoid colliding with the global `Map`